platform/linux/mbedtls/network_mbedtls_wrapper.c

```diff
@@ -286,73 +286,116 @@ IoT_Error_t iot_tls_connect(Network *pNetwork, TLSConnectParams *params) {
+				return NETWORK_SSL_WRITE_ERROR;
 			}
-		}
-		if(isErrorFlag) {
-			break;
-		}
-	}
 
-	*written_len = written_so_far;
+			/* Successfully sent data, so reset the timeout */
+			init_timer(&writeTimer);
+			countdown_ms(&writeTimer, IOT_SSL_WRITE_RETRY_TIMEOUT_MS);
 
-	if(isErrorFlag) {
-		return NETWORK_SSL_WRITE_ERROR;
-	} else if(has_timer_expired(timer) && written_so_far != len) {
-		return NETWORK_SSL_WRITE_TIMEOUT_ERROR;
+			txLen += ret;
+			pMsg += ret;
+			len -= ret;
+		} else if(ret == MBEDTLS_ERR_SSL_WANT_READ ||
+				ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			if(has_timer_expired(&writeTimer)) {
+				return NETWORK_SSL_WRITE_TIMEOUT_ERROR;
+			}
+		} else {
+			IOT_ERROR(" failed\n  ! mbedtls_ssl_write returned -0x%x\n\n", (unsigned int) -ret);
+			/* All other negative return values indicate connection needs to be reset.
+			 * Will be caught in ping request so ignored here */
+			return NETWORK_SSL_WRITE_ERROR;
+		}
 	}
 
+	*written_len = txLen;
 	return SUCCESS;
 }
 
+/* When this much time has elapsed after receiving MBEDTLS_ERR_SSL_WANT_READ,
+ * MBEDTLS_ERR_SSL_WANT_WRITE, or MBEDTLS_ERR_SSL_TIMEOUT, then iot_tls_read
+ * will return NETWORK_SSL_READ_TIMEOUT_ERROR. */
+#define IOT_SSL_READ_RETRY_TIMEOUT_MS 10
+
 IoT_Error_t iot_tls_read(Network *pNetwork, unsigned char *pMsg, size_t len, Timer *timer, size_t *read_len) {
-	mbedtls_ssl_context *ssl = &(pNetwork->tlsDataParams.ssl);
-	size_t rxLen = 0;
+	mbedtls_ssl_context *pSsl = &(pNetwork->tlsDataParams.ssl);
+	size_t rxLen = 0U;
 	int ret;
+	/* This timer checks for a timeout whenever MBEDTLS_ERR_SSL_WANT_READ,
+	 * MBEDTLS_ERR_SSL_WANT_WRITE, or MBEDTLS_ERR_SSL_TIMEOUT are returned by
+	 * mbedtls_ssl_read. Timeout is specified by IOT_SSL_READ_RETRY_TIMEOUT_MS. */
+	Timer readTimer;
+
+	/* This variable is unused */
+	(void) timer;
+
+	/* The timer must be started in case no bytes are read on the first try */
+	init_timer(&readTimer);
+	countdown_ms(&readTimer, IOT_SSL_READ_RETRY_TIMEOUT_MS);
+
+	while(len > 0U) {
+		/* This read will timeout after IOT_SSL_READ_TIMEOUT if there's no data to be read */
+		ret = mbedtls_ssl_read(pSsl, pMsg, len);
+
+		if(ret > 0) {
+			if((size_t) ret > len) {
+				IOT_ERROR("More bytes read than requested\n\n");
+				return NETWORK_SSL_WRITE_ERROR;
+			}
+
+			/* Successfully received data, so reset the timeout */
+			init_timer(&readTimer);
+			countdown_ms(&readTimer, IOT_SSL_READ_RETRY_TIMEOUT_MS);
 
-	while (len > 0) {
-		// This read will timeout after IOT_SSL_READ_TIMEOUT if there's no data to be read
-		ret = mbedtls_ssl_read(ssl, pMsg, len);
-		if (ret > 0) {
 			rxLen += ret;
 			pMsg += ret;
 			len -= ret;
-		} else if (ret == 0 || (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE && ret != MBEDTLS_ERR_SSL_TIMEOUT)) {
+		} else if(ret == MBEDTLS_ERR_SSL_WANT_READ || 
+				ret == MBEDTLS_ERR_SSL_WANT_WRITE ||
+				ret == MBEDTLS_ERR_SSL_TIMEOUT) {
+			if(has_timer_expired(&readTimer)) {
```

Comment:
is the indentation correct here?

Reply:
I don't like it either, but this file exclusively uses tabs... I really want to uncrustify all files in V3.